geomstats/geometry/poincare_ball.py

```diff
@@ -317,3 +317,268 @@ def inner_product_matrix(self, base_point=None):
+        as_o_gs = (1 + gs.erf(ins_gs)) * gs.exp(ins_squared_gs)
+        bs_o_gs = binomial_coefficient * as_o_gs
+        r_gs = alternate_neg * bs_o_gs
+
+        norm_func = NORMALIZATION_FACTOR_CST * variances * \
+            r_gs.sum(0) * (1 / (2 ** (dimension - 1)))
+
+        return norm_func
+
+    @classmethod
+    def _compute_alpha(cls, dim, current_dim):
+        """Compute factor used in normalisation factor.
+
+        Compute alpha factor given the two arguments.
+        """
+        return (dim - 1 - 2 * current_dim) / SQRT_2
+
+    def norm_factor_gradient(self, variances, dimension):
+        """Compute normalisation factor and its gradient.
+
+        Compute normalisation factor given current variance
+        and dimensionality.
+
+        Parameters
+        ----------
+        variances : array-like, shape=[n]
+            Value of variance.
+
+        dimension : int
+            Dimension of the space
+
+        Returns
+        -------
+        norm_factor : array-like, shape=[n]
+            Normalisation factor
+        norm_factor_gradient : array-like, shape=[n]
+            Gradient of the normalisation factor
+
+        """
+        variances = gs.transpose(gs.to_ndarray(variances, to_ndim=2))
+        dim_range = gs.arange(0, dimension, 1.)
+        alpha = self._compute_alpha(dimension, dim_range)
+
+        binomial_coefficient = gs.ones(dimension)
+        binomial_coefficient[1:] = \
+            (dimension - 1 + 1 - dim_range[1:]) / dim_range[1:]
+        binomial_coefficient = gs.cumprod(binomial_coefficient)
+
+        beta = ((-gs.ones(dimension)) ** dim_range) * binomial_coefficient
+
+        sigma_repeated = gs.repeat(variances, dimension, -1)
+        prod_alpha_sigma = gs.einsum('ij,j->ij', sigma_repeated, alpha)
+        term_2 = \
+            gs.exp((prod_alpha_sigma) ** 2) * (1 + gs.erf(prod_alpha_sigma))
+        term_1 = gs.sqrt(gs.pi / 2.) * (1. / (2 ** (dimension - 1)))
+        term_2 = gs.einsum('ij,j->ij', term_2, beta)
+        norm_factor = \
+            term_1 * variances * gs.sum(term_2, axis=-1, keepdims=True)
+        grad_term_1 = 1 / variances
+
+        grad_term_21 = 1 / gs.sum(term_2, axis=-1, keepdims=True)
+
+        grad_term_211 = \
+            gs.exp((prod_alpha_sigma) ** 2) \
+            * (1 + gs.erf(prod_alpha_sigma)) \
+            * gs.einsum('ij,j->ij', sigma_repeated, alpha ** 2) * 2
+
+        grad_term_212 = gs.repeat(gs.expand_dims((2 / gs.sqrt(gs.pi))
+                                                 * alpha, axis=0),
+                                  variances.shape[0], axis=0)
+
+        grad_term_22 = grad_term_211 + grad_term_212
+        grad_term_22 = gs.einsum("ij, j->ij", grad_term_22, beta)
+        grad_term_22 = gs.sum(grad_term_22, axis=-1, keepdims=True)
+
+        norm_factor_gradient = grad_term_1 + (grad_term_21 * grad_term_22)
+
+        return gs.squeeze(norm_factor), gs.squeeze(norm_factor_gradient)
```

Comment:
While it packs the code together nicely, I wonder if a class is the best choice for the normalization factor? 

Some functions are static, it seems that the attributes only store intermediary computations, there is a method called `normalization_factor`, which is also the name of the class itself.

Why did you choose a class, as opposed to - for example - a function calling auxiliary functions?

Reply:
great suggestion, I removed the class and added all the functions as part of PoincareBallMetric.